File: model/aether_model/phy/tx.py
"""Frame transmitter: header + coded symbols → complex baseband → audio (roadmap P1-4).

A frame is ``[SC, SC] + data symbols`` (the SC sequence encodes the frame type); data
symbol ``i`` is a full pilot symbol when ``i`` is in ``layout.pilot_symbol_indices`` — in
DATA frames its data carriers carry the mode's PN chips — and otherwise carries the next
``n_data_carriers`` constellation symbols (time-major) on its data carriers with comb
pilots. Frames are separated by ``guard_symbols`` of silence so the windowed edges decay
cleanly; the trailing taper of one frame overlap-adds into the guard.
"""

from __future__ import annotations

import numpy as np
from numpy.typing import NDArray

from aether_model.frame.modes import MODES, FrameLayout
from aether_model.phy.ofdm import OfdmModulator
from aether_model.phy.papr import ClipAndFilter, clip_target_db
from aether_model.phy.passband import BasebandToAudio
from aether_model.phy.preamble import FrameHeader, FrameType, preamble
from aether_model.waveform import WIDE_2300, WaveformParams

ComplexArray = NDArray[np.complex128]


class FrameTransmitter:
    def __init__(self, params: WaveformParams = WIDE_2300, papr_reduction: bool = True) -> None:
        self.p = params
        self.mod = OfdmModulator(params)
        self.pre = preamble(params)
        self.n_data = len(self.mod.cmap.data_carriers)
        self.papr_reduction = papr_reduction
        """Clip-and-filter the finished burst (ADR-0004). Off reproduces the raw OFDM
        envelope, for comparisons and for a PA that is already linear enough."""
        self._clippers: dict[float, ClipAndFilter] = {}

    def _clipper(self, target_db: float) -> ClipAndFilter:
        if target_db not in self._clippers:
            self._clippers[target_db] = ClipAndFilter(self.p, target_papr_db=target_db)
        return self._clippers[target_db]

    def symbol_values(
        self, header: FrameHeader, layout: FrameLayout, qam: ComplexArray
    ) -> list[ComplexArray]:
        """Carrier-value vectors for every OFDM symbol of the frame, preamble first."""
        qam = np.asarray(qam, dtype=np.complex128)
        if qam.shape != (layout.qam_symbols,):
            raise ValueError(
                f"expected {layout.qam_symbols} constellation symbols, got {qam.shape}"
            )
        symbols = self.pre.symbols(header)
        pilots = set(layout.pilot_symbol_indices)
        pos = 0
        pilot_no = 0
        for i in range(layout.data_symbols):
            if i in pilots:
                chips = None
                if header.frame_type is FrameType.DATA:
                    chips = self.pre.mode_chips(header.mode, pilot_no, header.rv)
                pilot_no += 1
                symbols.append(self.mod.symbol_values(None, full_pilot=True, chips=chips))
            else:
                symbols.append(self.mod.symbol_values(qam[pos : pos + self.n_data]))
                pos += self.n_data
        assert pos == layout.qam_symbols
        return symbols

    def baseband(self, header: FrameHeader, layout: FrameLayout, qam: ComplexArray) -> ComplexArray:
        """Windowed complex-baseband waveform of one frame (``layout.samples + taper`` long),
        peak-reduced per ADR-0004 unless ``papr_reduction`` is off."""
        x = self.mod.modulate(self.symbol_values(header, layout, qam))
        if not self.papr_reduction:
            return x
        target = clip_target_db(MODES[header.mode].modulation.bits_per_symbol)
        return self._clipper(target).process(x)

    def burst(
        self,
        header: FrameHeader,
        layout: FrameLayout,
        qam: ComplexArray,
        lead_silence_s: float = 0.0,
        tail_silence_s: float = 0.0,
    ) -> ComplexArray:
        """Frame with silence padding at the baseband rate (for PTT lead/tail or tests)."""
        fs = self.p.fs_baseband
        lead = np.zeros(round(lead_silence_s * fs), dtype=np.complex128)
        tail = np.zeros(round(tail_silence_s * fs), dtype=np.complex128)
        return np.concatenate((lead, self.baseband(header, layout, qam), tail))

    def audio(self, baseband: ComplexArray) -> NDArray[np.float32]:
        """Convert a complete baseband burst to 48 kHz audio (fresh converter per burst)."""
        conv = BasebandToAudio(self.p)
        # flush the converter's group delay so the burst is complete
        pad = np.zeros(2 * conv.tx_delay_samples // self.p.resample_factor + 8, dtype=np.complex128)
        return conv.process(np.concatenate((baseband, pad)))

File: core/aether-phy/src/tx.rs
//! Frame transmitter: a header and coded symbols into complex baseband.
//!
//! A frame is `[SC, SC]` followed by the layout's data symbols. Data symbol `i` is a full
//! pilot symbol when `i` is one of the layout's pilot symbol indices — and in a DATA frame
//! its data carriers then carry the mode and redundancy version as PN chips — and otherwise
//! carries the next `n_data_carriers` constellation symbols, time-major, alongside its comb
//! pilots.
//!
//! Peak reduction (ADR-0004) is *not* applied here yet; see the crate documentation for what
//! that means for comparisons against the reference model.

use crate::{
    constellation::Complex,
    modes::FrameLayout,
    ofdm::OfdmModulator,
    preamble::{FrameHeader, FrameType, Preamble},
    waveform::{WIDE_2300, WaveformParams},
};

/// Assembles frames.
#[derive(Debug)]
pub struct FrameTransmitter {
    params: WaveformParams,
    modulator: OfdmModulator,
    preamble: Preamble,
    n_data_carriers: usize,
}

/// Why a frame could not be assembled.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TxError {
    /// The wrong number of constellation symbols for the layout.
    BadSymbolCount {
        /// What the layout needs.
        expected: usize,
        /// What arrived.
        got: usize,
    },
}

impl core::fmt::Display for TxError {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::BadSymbolCount { expected, got } => {
                write!(f, "expected {expected} constellation symbols, got {got}")
            }
        }
    }
}

impl core::error::Error for TxError {}

impl FrameTransmitter {
    /// Build the transmitter.
    #[must_use]
    pub fn new(params: WaveformParams) -> Self {
        let modulator = OfdmModulator::new(params);
        let n_data_carriers = modulator.map().data_carriers().len();
        Self {
            params,
            modulator,
            preamble: Preamble::new(params),
            n_data_carriers,
        }
    }

    /// The modulator, for callers that need the carrier map.
    #[must_use]
    pub fn modulator(&self) -> &OfdmModulator {
        &self.modulator
    }

    /// Carrier values for every OFDM symbol of the frame, preamble first.
    ///
    /// # Errors
    /// If `qam` is not exactly the layout's slot count.
    pub fn symbol_values(
        &self,
        header: &FrameHeader,
        layout: &FrameLayout,
        qam: &[Complex],
    ) -> Result<Vec<Vec<Complex>>, TxError> {
        let expected = layout.qam_symbols();
        if qam.len() != expected {
            return Err(TxError::BadSymbolCount {
                expected,
                got: qam.len(),
            });
        }
        let mut symbols = self.preamble.symbols(header);
        let pilots = layout.pilot_symbol_indices();
        let mut position = 0usize;
        let mut pilot_number = 0usize;
        for index in 0..layout.data_symbols {
            if pilots.contains(&index) {
                let chips = match header.frame_type {
                    FrameType::Data => Some(self.preamble.mode_chips(
                        header.mode,
                        pilot_number,
                        header.rv,
                    )),
                    FrameType::Control => None,
                };
                pilot_number += 1;
                symbols.push(self.modulator.pilot_symbol(chips.as_deref()));
            } else {
                let slice = &qam[position..position + self.n_data_carriers];
                position += self.n_data_carriers;
                symbols.push(self.modulator.data_symbol(slice));
            }
        }
        debug_assert_eq!(position, expected);
        Ok(symbols)
    }

    /// Windowed complex-baseband waveform of one frame, `layout.samples() + taper` long.
    ///
    /// # Errors
    /// If `qam` is not exactly the layout's slot count.
    pub fn baseband(
        &self,
        header: &FrameHeader,
        layout: &FrameLayout,
        qam: &[Complex],
    ) -> Result<Vec<Complex>, TxError> {
        let symbols = self.symbol_values(header, layout, qam)?;
        Ok(self.modulator.modulate(&symbols))
    }

    /// The numerology in use.
    #[must_use]
    pub fn params(&self) -> WaveformParams {
        self.params
    }
}

impl Default for FrameTransmitter {
    fn default() -> Self {
        Self::new(WIDE_2300)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{
        codec::FrameCodec,
        modes::{CONTROL_MODE, LONG, MODES, SHORT},
        ofdm::OfdmDemodulator,
    };

    fn payload(n: usize) -> Vec<u8> {
        (0..n).map(|i| ((i * 29) % 256) as u8).collect()
    }

    #[test]
    fn a_frame_is_the_expected_length() {
        let tx = FrameTransmitter::default();
        let codec = FrameCodec::new(MODES[4], LONG).unwrap();
        let qam = codec.encode(&payload(codec.payload_bytes), 0).unwrap();
        let header = FrameHeader::new(FrameType::Data, 4, 0).unwrap();
        let waveform = tx.baseband(&header, &LONG, &qam).unwrap();
        assert_eq!(waveform.len(), LONG.samples() + WIDE_2300.taper_samples);
    }

    #[test]
    fn the_frame_has_the_right_symbol_structure() {
        let tx = FrameTransmitter::default();
        let codec = FrameCodec::new(MODES[8], LONG).unwrap();
        let qam = codec.encode(&payload(codec.payload_bytes), 2).unwrap();
        let header = FrameHeader::new(FrameType::Data, 8, 2).unwrap();
        let symbols = tx.symbol_values(&header, &LONG, &qam).unwrap();
        assert_eq!(symbols.len(), LONG.total_symbols());
        assert_eq!(
            symbols[0], symbols[1],
            "the two preamble symbols are identical"
        );
        // every symbol carries one value per active carrier
        assert!(symbols.iter().all(|s| s.len() == WIDE_2300.n_carriers()));
    }

    #[test]
    fn demodulating_a_frame_returns_the_transmitted_carrier_values() {
        let tx = FrameTransmitter::default();
        let demodulator = OfdmDemodulator::new(WIDE_2300);
        let codec = FrameCodec::new(MODES[4], LONG).unwrap();
        let qam = codec.encode(&payload(codec.payload_bytes), 0).unwrap();
        let header = FrameHeader::new(FrameType::Data, 4, 0).unwrap();
        let expected = tx.symbol_values(&header, &LONG, &qam).unwrap();
        let waveform = tx.baseband(&header, &LONG, &qam).unwrap();
        let period = WIDE_2300.symbol_samples();

        // the last symbol has no successor to overlap with, so stop one short
        for (index, want) in expected.iter().enumerate().take(LONG.total_symbols() - 1) {
            let got = demodulator
                .carriers(&waveform, index * period)
                .expect("in range");
            for (carrier, (g, w)) in got.iter().zip(want).enumerate() {
                assert!(
                    (g.0 - w.0).abs() < 1e-9 && (g.1 - w.1).abs() < 1e-9,
                    "symbol {index}, carrier {carrier}"
                );
            }
        }
    }

    #[test]
    fn a_control_frame_carries_no_chips() {
        let tx = FrameTransmitter::default();
        let codec = FrameCodec::new(CONTROL_MODE, SHORT).unwrap();
        let qam = codec.encode(&payload(codec.payload_bytes), 0).unwrap();
        let symbols = tx
            .symbol_values(&FrameHeader::control(), &SHORT, &qam)
            .unwrap();
        let map = tx.modulator().map();
        // a control frame's pilot symbols carry the plain pilot sequence on every carrier
        let pilot_symbol = &symbols[2]; // first data symbol, which is a pilot symbol
        for &carrier in map.data_carriers() {
            let expected = map.pilot_sequence()[carrier];
            assert!((pilot_symbol[carrier].0 - expected.0).abs() < 1e-12);
            assert!((pilot_symbol[carrier].1 - expected.1).abs() < 1e-12);
        }
    }

    #[test]
    fn a_data_frame_puts_its_chips_on_the_pilot_symbols() {
        let tx = FrameTransmitter::default();
        let codec = FrameCodec::new(MODES[6], LONG).unwrap();
        let qam = codec.encode(&payload(codec.payload_bytes), 3).unwrap();
        let header = FrameHeader::new(FrameType::Data, 6, 3).unwrap();
        let symbols = tx.symbol_values(&header, &LONG, &qam).unwrap();
        let map = tx.modulator().map();
        let preamble = Preamble::default();
        for (pilot_number, &index) in LONG.pilot_symbol_indices().iter().enumerate() {
            let symbol = &symbols[2 + index];
            let chips = preamble.mode_chips(6, pilot_number, 3);
            for (&carrier, &chip) in map.data_carriers().iter().zip(&chips) {
                assert!(
                    (symbol[carrier].0 - chip).abs() < 1e-12,
                    "pilot symbol {index}"
                );
                assert!(symbol[carrier].1.abs() < 1e-12);
            }
        }
    }

    #[test]
    fn the_wrong_symbol_count_is_an_error() {
        let tx = FrameTransmitter::default();
        let header = FrameHeader::new(FrameType::Data, 0, 0).unwrap();
        assert!(matches!(
            tx.baseband(&header, &LONG, &[(0.0, 0.0); 3]),
            Err(TxError::BadSymbolCount { .. })
        ));
    }
}
